Add best practices (#178)
near/near-sdk-rs: HELP.md

# Contract best practices

## Init once

Make sure you check that state doesn't exists in `init` using

```rust

> assert!(!env::state_exists(), "The contract is already initialized");

Reviewer: Should this become an argument on the `init` marco? e.g. `init(once)`?

Author: Or `#[init_once]` https://github.com/near/near-sdk-rs/issues/95

prevent default state initialization.

```rust
impl Default for Contract {
    fn default() -> Self {
        env::panic(b"The contract should be initialized before usage")
    }
}
```

## Public vs private methods

`pub fn` makes a method public and exposes it in a contract. It means anyone can call it.
`fn` makes the method private and it's not exposed from the contract. No one can call it directly. It can only be called
within a contract directly (not through a promise).

## Private callbacks

If you're using callbacks, makes sure you check the caller.

```rust
assert_eq!(env::current_account_id(), env::predecessor_account_id(), "Can only be called within a contract");
```

Callbacks have to be public methods.

## JSON types

NEAR currently expects contracts to support JSON serialization. JSON can't handle large integers (above 2**53 bits).
That's why you should use helper classes from the `json_types` in `near_sdk` for `u64` and `u128`.
We provide types `U64` and `U128`, that wraps the integer into a struct and implements JSON serialization and
deserialization as a base-10 strings.

E.g.

```rust
pub fn add(&self, a: U128, b: U128) -> U128 {

>     (a.0 + b.0).into()

Reviewer: I found this to be annoying.  Is there really not a way to handle this with the bindgen?

Author: Not easily. We don't want to make it a requirement to serialize long numbers as strings. Hopefully JSON standard can improve in the future and we can use regular long numbers with string quotes.

Author: We can implement math on `U128`, so you'll be able to do simply `a + b`

Reviewer: Yeah that's a good idea.

}
```

## View vs Change method

By default, `near_sdk` assumes that the method is a `view` if it has `&self` and method is `change` if it has `&mut self`.
View methods usually don't have ability to change state and have limited context.

> Change methods will save the modified state back to the contract.

Reviewer: I would also mention that you can modify storage the contract's state.

## `STATE` storage key

The contract will save the main structure state under the storage key `STATE`.
Make sure you don't modify it e.g. through collection prefixes or raw storage access.

## Enable overflow checks

It's usually helpful to panic on integer overflow. To enable it, add the following into your `Cargo.toml` file:

```toml
[profile.release]
overflow-checks = true
```

## Use `assert!` early in the method.

Try to validate the input, context, state and access first before making any actions. This will safe gas for the caller to fail earlier.

## Use `env::log`

Use logging for debugging and notifying user.
When you need a formatted message, you can use the following:

```rust

> env::log(format!("Transferred {} tokens from {} to {}", amount, sender_id, receiver_id).as_bytes());

Reviewer: I've always found this API strange.  Why not just make log a macro so that you don't have to `.as_bytes` every time?

Author: Good idea, please add an issue for this.

```

## Return `Promise`

If your method makes a cross-contract call, you may want to return the `Promise` that the contract creates.
It allows the caller to wait for the result of the promise instead of the returning the result immediately.
Because if the promise fails for some reason, the caller will not know about this or the caller may display the result earlier.

E.g.

```rust
pub fn withdraw(&mut self, amount: U128) -> Promise {

>     Promise::new(self.owner_id.clone()).transfer(amount.0)

Reviewer: Perhaps we should also change the promise API so that it uses `U128`

Author: No, it's good for now. I still have a hope that we'll use borsh instead of JSON. Inner numbers don't have to be wrapped in `U128`

Reviewer: Do you mean use borsh for input serialization too?

Change methods (see below) are serializing the main contract structure at the end and stores the new value into the storage.

Persistent collection helps store extra data in the persistent storage outside of main structure.
NEAR SDK provides the following collections:
- `Vector` - An iterable implementation of vector.
- `LookupMap` - An non-iterable implementation of a map.
- `LookupSet` - An non-iterable implementation of a set.
- `UnorderedMap` - An iterable implementation of a map.
- `UnorderedSet` - An iterable implementation of a set.
- `TreeMap` - An iterable sorted map based on AVL-tree
- `LazyOption` - An `Option` for a single value.

Every instance of a persistent collection requires a unique storage prefix.
The prefix is used to generate internal keys to store data in the persistent storage.
These internal keys has to be unique to avoid collisions (even with key `STATE`).

## Generating unique prefixes for persistent collections

When a contract gets complicated, there may be multiple different
collections that may not be all part of the main structure, but instead be part of sub-structure or nested collections.
They all need to have unique prefixes.

> #### The traditional way

Reviewer 2: should we make this a small stub and link to appendix of doc just to keep the flow and emphasis on "the new way"?

Traditionally, we hardcode them in the constructor using a short one letter prefix that was converted to a vec.
When used nested collection, we had to manually construct a prefix.

```rust
#[near_bindgen]
impl Contract {
    #[init]
    pub fn new() -> Self {
        Self {
            accounts: UnorderedMap::new(b"a".to_vec()),
            tokens: LookupMap::new(b"t".to_vec()),
            metadata: LazyOption::new(b"m".to_vec()),
        }
    }

    fn get_tokens(&self, account_id: &AccountId) -> UnorderedSet<String> {
        let tokens = self.accounts.get(account_id).unwrap_or_else(|| {
            // Constructing a unique prefix for a nested UnorderedSet.
            let mut prefix = Vec::with_capacity(33);
            // Adding unique prefix.
            prefix.push(b's');
            // Adding the hash of the account_id (key of the outer map) to the prefix.
            // This is needed to differentiate across accounts.
            prefix.extend(env::sha256(account_id.as_bytes()));
            UnorderedSet::new(prefix)
        });
        tokens
    }
}
```

> #### The new way

Reviewer 2: like

- `Vector` - An iterable implementation of vector.
- `LookupMap` - An non-iterable implementation of a map.
- `LookupSet` - An non-iterable implementation of a set.
- `UnorderedMap` - An iterable implementation of a map.
- `UnorderedSet` - An iterable implementation of a set.
- `TreeMap` - An iterable sorted map based on AVL-tree
- `LazyOption` - An `Option` for a single value.

Every instance of a persistent collection requires a unique storage prefix.
The prefix is used to generate internal keys to store data in the persistent storage.
These internal keys has to be unique to avoid collisions (even with key `STATE`).

## Generating unique prefixes for persistent collections

When a contract gets complicated, there may be multiple different
collections that may not be all part of the main structure, but instead be part of sub-structure or nested collections.
They all need to have unique prefixes.

We can introduce an `enum` for tracking storage prefixes and keys.
And then use borsh serialization to construct a unique prefix for every collection.
It's as efficient as manually constructing them, because Borsh enum only takes one byte.

```rust

> #[derive(BorshSerialize)]

Reviewer 3: Why no need for `BorshDeserialize` here?

Author: We're only serializing keys, but not deserializing them.

pub enum StorageKeys {
    Accounts,
    SubAccount { account_hash: Vec<u8> },
    Tokens,
    Metadata,
}

#[near_bindgen]
impl Contract {
    #[init]
    pub fn new() -> Self {
        Self {
            accounts: UnorderedMap::new(StorageKeys::Accounts.try_to_vec().unwrap()),
            tokens: LookupMap::new(StorageKeys::Tokens.try_to_vec().unwrap()),
            metadata: LazyOption::new(StorageKeys::Metadata.try_to_vec().unwrap()),
        }
    }
    
    fn get_tokens(&self, account_id: &AccountId) -> UnorderedSet<String> {
        let tokens = self.accounts.get(account_id).unwrap_or_else(|| {
            UnorderedSet::new(
                StorageKeys::SubAccount { account_hash: env::sha256(account_id.as_bytes()) }
                    .try_to_vec()

>                     .unwrap(),

Reviewer 3: is there a way to avoid all the `.try_to_vec().unwrap()`? It really clutters up a great interface.

Author: we can add `to_key` impl that does this:
```rust
impl StorageKeys {
  pub fn to_vec(&self) -> Vec<u8> {
    self.try_to_vec().unwrap()
  }
}
```

Author: We can also change `UnorderedSet::new` to take `IntoVec` and then it may simplify the code.

Reviewer 3: Would we have to do that with every kind of collection? Is it worth it? What would this look like, then? This?

```rust
StorageKeys::SubAccount { account_hash: env::sha256(account_id.as_bytes()) }
    into(),
```

It will panic if the old state (under key `STATE`) is present in the storage.

But if you need to re-initialize the contract STATE, you need to use `#[init(ignore_state)]` instead.
This will NOT check that the state exists and you can use it in case you need to upgrade contract and migrate state.

```rust
#[near_bindgen]
impl Contract {
    #[init(ignore_state)]
    pub fn migrate_state(new_data: String) -> Self {
        // Deserialize the state using the old contract structure.
        let old_contract: OldContract = env::state_read().expect("Old state doesn't exist");
        // Verify that the migration can only be done by the owner.
        // This is not necessarily, if the upgrade is done internally.
        assert_eq!(
            &env::predecessor_account_id(),
            &old_contract.owner_id,
            "Can only be called by the owner"
        );

        // Create the new contract using the data from the old contract.
        Self { owner_id: old_contract.owner_id, data: old_contract.data, new_data }
    }
}
```

## Use `PanicOnDefault`

By default `near_sdk` allows a contract to be initialized with the default state.

> Usually, if you have an initializer, you may want to prevent it.

Reviewer 3: We should say _why_ here. Or give an example of something bad that could happen if you forget this.

        env::log(b"This is a callback");
    }
}
```

## Integer JSON types

NEAR Protocol currently expects contracts to support JSON serialization. JSON can't handle large integers (above 2**53 bits).
That's why you should use helper classes from the `json_types` in `near_sdk` for `u64` and `u128`.
We provide types `U64` and `U128`, that wraps the integer into a struct and implements JSON serialization and
deserialization as a base-10 strings.

You can convert from `U64` to `u64` and back using `std::convert::Into`, e.g.

```rust
#[near_bindgen]
impl Contract {
    pub fn mult(&self, a: U64, b: U64) -> U128 {
        let a: u64 = a.into();
        let b: u64 = b.into();
        let product = u128::from(a) * u128::from(b);
        product.into()
    }
}
```

You can also access inner values and using `.0` and `U128(5)`, e.g.

```rust
#[near_bindgen]
impl Contract {
    pub fn sum(&self, a: U128, b: U128) -> U128 {
        U128(a.0 + b.0)
    }
}
```

## `Base64VecU8` JSON type

Often the contract needs to receive or return a binary data.
Encoding a `Vec<u8>` with JSON will lead to an integer array, e.g. `[110, 101, 97, 114]`
This is inefficient in both compute and space.

> `Base64VecU8` is a wrapper on top of `Vec<u8>` that allows to pass it as arguments or result.

Reviewer 3: Why is this more efficient in either compute or space? 

We store it internally as `Vec<u8>` anyhow, so the space consideration seems irrelevant.

And for compute on in/out, base64 is [about 33% larger than binary](https://stackoverflow.com/questions/33811673/why-the-size-of-base64-encoded-string-is-larger-than-the-original-file). How then is serializing & deserializing base64 more compute-efficient than using binary?

If neither of these are true, is there some other situation that should make people reach for `Base64VecU8`?

```rust
#[near_bindgen]
#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
pub struct Contract {
    // Notice, internally we store `Vec<u8>` 
    pub data: Vec<u8>,
}

#[near_bindgen]
impl Contract {
    #[init]
    pub fn new(data: Base64VecU8) -> Self {
        Self {
            data: data.into(),
        }
    }

    pub fn get_data(self) -> Base64VecU8 {
        self.data.into()
    }
}
```

## View vs Change method

`near_sdk` assumes that the method is a `view` if it uses `&self` or `self` and method is `change` if it has `&mut self`.

View methods don't save the contract STATE at the end of the method execution.

Change methods will save the modified STATE at the end of the method execution. They can also modify the state in persistent collections.

Note: Change methods will also check that the function call doesn't have attached deposit, unless the method is marked with the `#[payable]` macro.

```rust
#[near_bindgen]
impl Contract {
    /// View method. Requires cloning the account id.
    pub fn get_owner_id(&self) -> AccountId {
        self.owner_id.clone()
    }

>     /// View method. More efficient, but can't be reused internally, because it consumes self.

Reviewer 3: Is there a reason to prefer Rust's doc comments with three slashes to non-doc comments with only two?

Reviewer 4: There's a good example here, and I think this is typically helpful as it is here right above a function. A non-doc comment might be in the implementation if something seems unintuitive and it's helpful to add a note.
https://doc.rust-lang.org/reference/comments.html#doc-comments

Reviewer 3: Ok, but the docs are not really _used_ yet, in the NEAR ecosystem. 

I'm guessing that someday the Metadata API might use doc comments.

Reviewer 4: I don't think that's accurate, we just haven't fully fleshed it out yet, which I hope we'll do this quarter.
<img width="2089" alt="Screen Shot 2021-04-09 at 2 02 01 PM" src="https://user-images.githubusercontent.com/1042667/114240374-43d43400-993c-11eb-80d1-a33a7c9e7b31.png">

        let calculator_account_id: AccountId = CALCULATOR_ACCOUNT_ID.to_string();
        ext_calculator::sum(a, b, &calculator_account_id, NO_DEPOSIT, BASE_GAS)
    }
}
```

## Reuse crates from `near-sdk`

`near-sdk` re-exports the following crates:

- `borsh`
- `base64`
- `bs58`
- `serde`
- `serde_json`
- `wee_alloc` (Though you will likely use the `setup_alloc` macro instead of importing it directly)

Most common crates include `borsh` that is needed for internal STATE serialization and
`serde` for the external JSON serialization.

When marking structs with `serde::Serialize` you need to use `#[serde(crate = "near_sdk::serde")]`
to point serde into the correct base crate.

```rust
/// Import `borsh` from `near_sdk` crate 
use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
/// Import `serde` from `near_sdk` crate 
use near_sdk::serde::{Serialize, Deserialize};

/// Main contract structure serialized with Borsh
#[near_bindgen]
#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
pub struct Contract {
    pub pair: Pair,
}

/// Implements both `serde` and `borsh` serialization.
#[derive(Serialize, Deserialize, BorshDeserialize, BorshSerialize)]

> #[serde(crate = "near_sdk::serde")]

Reviewer 3: Is it common to implement both? If this is not common, can we change the example to something that is? If it is common, can we explain why it's useful?

Reviewer 4: added a comment about serde being useful

Reviewer 3: Great info about serde, thanks.

I'm also curious about using both "Serialize/Deserialize" and "BorshSerialize/BorshDeserialize."

Reviewer 4: Sure, feel free to make an issue. Here are some helpful links:
- https://borsh.io/
- https://docs.near.org/docs/concepts/storage-staking#use-a-binary-serialization-format-rather-than-json

pub struct Pair {
    pub a: u32,
    pub b: u32,
}

#[near_bindgen]
impl Contract {
    #[init]
    pub fn new(pair: Pair) -> Self {
        Self {
            pair,
        }
    }

    pub fn get_pair(self) -> Pair {
        self.pair
    }
}
```

## Use `setup_alloc!`

> The SDK provides a helper macro to setup a global allocator from `wee_alloc` crate:

Reviewer 3: Can we add a sentence or a link about what a global allocator is and why this is needed? 

Also, is there ever a situation when this is not needed, or when a contract author may prefer a different allocator? If not, why even make people include this boilerplate?

Reviewer 4: pushed link to global allocators

        let keys = self.status_updates.keys_as_vector();
        let values = self.status_updates.values_as_vector();
        (from_index..std::cmp::min(from_index + limit, self.status_updates.len()))
            .map(|index| (keys.get(index).unwrap(), values.get(index).unwrap()))
            .collect()
    }
}
```

## `LookupMap` vs `UnorderedMap`

### Functionality

- `UnorderedMap` supports iteration over keys and values, and also supports pagination. Internally, it has the following structures:
  - a map from a key to an index
  - a vector of keys
  - a vector of values
- `LookupMap` only has a map from a key to a value. Without a vector of keys, it doesn't have the ability to iterate over keys.

### Performance

`LookupMap` has a better performance and stores less data comparing to the `UnorderedMap`.

- `UnorderedMap` requires `2` storage reads to get the value and `3` storage writes to insert a new entry.
- `LookupMap` requires only one storage read to get the value and only one storage write to store it.

### Storage space

`UnorderedMap` requires more storage for an entry comparing to a `LookupMap`.

> - `UnorderedMap` stores the key twice (once in the first map and once in the vector of keys) and value once. It also has higher constant for storing the length of vectors and prefixes.

Reviewer 3: "higher constant"? Seems like maybe there's a missing word?

Reviewer 4: Looking at the `fn insert_raw` from `LookupMap` and `UnorderedMap` seems like the latter has to incorportate `u64` storage. I'll add a word or two here for readability.

- `LookupMap` stores key and value once.

## `LazyOption`

It's a type of persistent collection that only stores a single value.
The goal is prevent contract from deserializing the given value until it's needed.
An example can be a large blob of metadata that is only needed when it's requested in a view call,
but not needed for the majority of contract operations.

It acts like an `Option` that can either hold a value or not and also requires a unique prefix (a key in this case)
like other persistent collections.

> Comparing to other collections, the `LazyOption` allows to initialize the value during the constructor.

Reviewer 3: Mixing grammar and substantive changes with this suggestion; not sure if I understand the substantive parts enough to convey them accurately

```suggestion
Compared to other collections, `LazyOption` only allows you to initialize the value during initialization.
```

    pub fn get_owner_id2(self) -> AccountId {
        self.owner_id
    }

    /// Change method. Changes the state, and then saves the new state internally.
    pub fn set_owner_id(&mut self, new_owner_id: ValidAccountId) {
        self.owner_id = new_owner_id.into();
    }
}
```

For more information about `&self` versus `self` see the [rust book](https://doc.rust-lang.org/stable/book/ch05-03-method-syntax.html?highlight=capture%20self#defining-methods)

## Payable methods

To mark a change method as a payable, you need add `#[payable]` macro decorator. This will allow this change method
to receive attached deposits. Otherwise, if a deposit is attached to non-payable change method, the method will panic.

```rust
#[near_bindgen]
impl Contract {
    #[payable]
    pub fn take_my_money(&mut self) {
        env::log(b"Thanks!");
    }

    pub fn do_not_take_my_money(&mut self) {
        env::log(b"Thanks!");
    }
}
```

This is equivalent to:

```rust
#[near_bindgen]
impl Contract {
    pub fn take_my_money(&mut self) {
        env::log(b"Thanks!");
    }

    pub fn do_not_take_my_money(&mut self) {

>         let error = format!("", ident.to_string());

Author: ```suggestion
```

```rust
#[near_bindgen]
impl Contract {
    #[init(ignore_state)]
    pub fn migrate_state(new_data: String) -> Self {
        // Deserialize the state using the old contract structure.
        let old_contract: OldContract = env::state_read().expect("Old state doesn't exist");
        // Verify that the migration can only be done by the owner.
        // This is not necessary, if the upgrade is done internally.
        assert_eq!(
            &env::predecessor_account_id(),
            &old_contract.owner_id,
            "Can only be called by the owner"
        );

        // Create the new contract using the data from the old contract.
        Self { owner_id: old_contract.owner_id, data: old_contract.data, new_data }
    }
}
```

## Use `PanicOnDefault`

By default `near_sdk` allows a contract to be initialized with default state.
Usually, if you have an initializer, you will want to prevent this.
There is a helper derive macro `PanicOnDefault` to do this, e.g.

```rust
#[near_bindgen]
#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
pub struct Contract {
    pub data: String,
}

> ```

Reviewer 5: Why do we need a separate proc macro for this? seems like `near_bindgen` can generate `Default` impl  which panics if ther's a method annotated with `#[init]`? Or maybe, better still, the code-gen can be changed to not rely on `Default` at all if there's init? 

It feels a bit that this is a work-around rather a best practice :)

        #[callback]
        sum: U128,
        caller: AccountId
    ) -> Account {
        let account = self.accounts.get(caller)
            .unwrap_or_panic("account {} not found", caller);
        account.total += sum.0;
        account
    }
}
```

## Reuse crates from `near-sdk`

`near-sdk` re-exports the following crates:

- `borsh`
- `base64`
- `bs58`
- `serde`
- `serde_json`
- `wee_alloc` (Though you will likely use the `setup_alloc` macro instead of importing it directly)

Most common crates include `borsh` which is needed for internal STATE serialization and
`serde` for external JSON serialization.

When marking structs with `serde::Serialize` you need to use `#[serde(crate = "near_sdk::serde")]`
to point serde to the correct base crate.

```rust
/// Import `borsh` from `near_sdk` crate 
use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
/// Import `serde` from `near_sdk` crate 
use near_sdk::serde::{Serialize, Deserialize};

/// Main contract structure serialized with Borsh
#[near_bindgen]
#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
pub struct Contract {
    pub pair: Pair,
}

/// Implements both `serde` and `borsh` serialization.

> /// `serde` is typically useful when returning a struct in JSON form for a frontend.

Reviewer: ```suggestion
/// `serde` is typically useful when returning a struct in JSON format for a frontend.
```
